rfcs/0001-restructure.md

```diff
@@ -0,0 +1,192 @@
+# Restructure to improve composition
+
+## Proposal
+
+The existing Python language buildpacks will be broken down into a set of
+buildpacks that, by themselves, provide more limited functionality, but in
+composition, fill the same feature set that the language family already
+provides.
+
+The current set of buildpacks will be reorganized into the following:
+
+![Proposed Structure](/rfcs/assets/0001-proposed.png)
+
+* `cpython`
+  * provides: `cpython`
+  * requires: none
+
+* `pip`
+  * provides: `pip`
+  * requires: `cpython` during `build`
+
+* `pip-install`
+  * provides: `site-packages`
+  * requires: `cpython` and `pip` during `build`
+
+* `pipenv`
+  * provides: `pipenv`
+  * requires: `pip` and `cpython` during `build`
```

Comment:
Is there no way to install/provide pipenv without requiring pip?
Is this because pipenv requires additional modules not provided by `cpython`?

I notice later in the doc you mention "`We can reuse existing ecosystem
tooling like pip to aid in the installation of packages using pipenv`". But I think if it's possible, It would be cleaner for a pipenv workflow to exist that is independent of pip.

Reply:
I think this should be possible, but would require that we have different dependencies than what is present in the existing buildpacks. The existing buildpacks have source code for `pip` and `pipenv` as dependencies. This means we need to use `python` or `pip` to install each. I believe there is a binary distribution concept called a `wheel` that we could use.

Reply:
I didn't know about wheels. Pipenv does publish wheels (https://pypi.org/project/pipenv/#files) but it looks like the conventional way to even install a wheel is through pip, though it looks like a zip you can unpack to disk. Regardless, it would be pretty cool if the buildpack doesn't have to bother about building binary from a source distribution.

Reply:
According to [the docs](https://pipenv.pypa.io/en/latest/install/#crude-installation-of-pipenv) you can run` curl https://raw.githubusercontent.com/pypa/pipenv/master/get-pipenv.py | python`

Reply:
I took a look at pipenv's `whl` and it seems to to be a source code distribution and not binary.
A useful [stackoverflow conversation about whl](https://stackoverflow.com/questions/63150946/souce-code-getting-packaged-in-python-wheel#answer-63157653)

Reply:
I broke open the `source` and `wheel` distributions and they are both just tarballs of python source code. I think it may be possible to install `pipenv` without `pip` but that seems hard and unnecessary.